src/api/runtime_api/block_builder.rs

```diff
@@ -0,0 +1,131 @@
+/*
+   Copyright 2024 Supercomputing Systems AG
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+	   http://www.apache.org/licenses/LICENSE-2.0
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+use super::{RuntimeApi, RuntimeApiClient};
+use crate::{api::Result, rpc::Request};
+use ac_primitives::{config::Config, UncheckedExtrinsicV4};
+#[cfg(not(feature = "sync-api"))]
+use alloc::boxed::Box;
+use alloc::{vec, vec::Vec};
+use sp_core::{Bytes, Encode};
+use sp_inherents::{CheckInherentsResult, InherentData};
+use sp_runtime::ApplyExtrinsicResult;
+
+#[maybe_async::maybe_async(?Send)]
+pub trait BlockBuilderApi: RuntimeApi {
+	type ApplyExtrinsicResult;
+	type Block;
+	type InherentData;
+	type CheckInherentsResult;
+	type Header;
+
+	/// Apply the given extrinsic.
+	async fn apply_extrinsic<Address, Call, Signature, SignedExtra>(
+		&self,
+		extrinsic: UncheckedExtrinsicV4<Address, Call, Signature, SignedExtra>,
+		at_block: Option<Self::Hash>,
+	) -> Result<Self::ApplyExtrinsicResult>
+	where
+		Address: Encode,
+		Call: Encode,
+		Signature: Encode,
+		SignedExtra: Encode;
+
+	/// Apply the given opaque extrinsic.
+	async fn apply_opaque_extrinsic(
+		&self,
+		extrinsic: Vec<u8>,
+		at_block: Option<Self::Hash>,
+	) -> Result<Self::ApplyExtrinsicResult>;
```

Comment:
We already have the same pattern in the rpc-api, offering strongly typed and weak typed (opaque/encoded) functions where necessary (mostly regarding `UncheckedExtrinsic`).  The reasoning behind that is, while with the opaque method the user has the full freedom of which (encoded) extrinsic to apply, but users that do not need this freedom should use the normal, strongly typed method to get errors during compilation.

And I disagree that the functions should match 1:1 to the interface of the Substrate one, because the api-client should provide convenience functions for the users. As long as it stays clear which Api is addressed in the end, why not offer more functionality?

Reply:
I'm not against extra functionality. But it'd be nice to separate the code for the two and have a naming scheme that doesn't give new names to standard substrate functions.